README.md
# Quick Interact

[Quick Interact](https://pub.dev/packages/quick_interact) is a Flutter package inspired by the Instagram quick send feature. It provides a simple and intuitive API
to create interactive widgets with ease, allowing you to show quick interactions for any widget with various
configurations.

<div align="center">
  <img src="https://github.com/onatcipli/quick_interact/assets/26159899/2db8e0d8-5cee-4e0b-853f-9485f6534b2f" alt="Screen Recording 2023-09-25 at 20 59 36" width="250"/>
  <img src="https://github.com/onatcipli/quick_interact/assets/26159899/df99f6d9-3e53-411c-ad73-2f0eca8a5661" alt="Screen Recording 2023-09-25 at 20 58 54" width="250"/>
  <img src="https://github.com/onatcipli/quick_interact/assets/26159899/551d465e-8478-4c76-bb47-eeca95149f10" alt="Screen Recording 2023-09-25 at 20 59 16" width="250"/>
  <img src="https://github.com/onatcipli/quick_interact/assets/26159899/445b2014-c763-40ae-878c-8673d3231c61" alt="Screen Recording 2023-09-25 at 21 00 45" width="250"/>
  <img src="https://github.com/onatcipli/quick_interact/assets/26159899/0ed01e0b-de16-4c74-a8c4-c1ef7265c144" alt="Screen Recording 2023-09-25 at 21 00 34" width="250"/>
  <img src="https://github.com/onatcipli/quick_interact/assets/26159899/6f85fe5e-90dc-42f7-9b34-d6c0f224ff9f" alt="Screen Recording 2023-09-25 at 21 00 16" width="250"/>
</div>


## Features

- Display quick interactions for any widget
- Customize the interaction with various configurations
- Easy to use and integrate into any Flutter project

## Installation

To use this package, add `quick_interact` as a dependency in your `pubspec.yaml` file.

```yaml
dependencies:
  quick_interact: ^latest_version
```

Or you can install it from the command line:

bash

```sh
flutter pub add quick_interact
```

## Usage

Here is a basic example of using Quick Interact:

```dart
Widget iconQuickReaction({required QuickInteractConfig config}) {
  return Center(
    child: QuickInteractions(
      quickInteractionWidgets: const [
        Icon(Icons.thumb_up),
        Icon(Icons.thumb_down),
        Icon(Icons.favorite),
        Icon(Icons.bookmark),
      ],
      onQuickInteractCompleted: (index) {
        print('Quick interaction selected: $index');
      },
      config: config,
      child: const Icon(Icons.send),
    ),
  );
}
```

### Example User Avatar Usage

```dart
Widget userAvatarQuickReaction({required QuickInteractConfig config}) {
  return Center(
    child: QuickInteractions(
      quickInteractionWidgets: const [
        CircleAvatar(
          radius: 15,
          backgroundImage:
          NetworkImage('https://randomuser.me/api/portraits/men/1.jpg'),
        ),
        CircleAvatar(
          radius: 15,
          backgroundImage:
          NetworkImage('https://randomuser.me/api/portraits/women/2.jpg'),
        ),
        CircleAvatar(
          radius: 15,
          backgroundImage:
          NetworkImage('https://randomuser.me/api/portraits/men/3.jpg'),
        ),
        CircleAvatar(
          radius: 15,
          backgroundImage:
          NetworkImage('https://randomuser.me/api/portraits/women/4.jpg'),
        ),
      ],
      onQuickInteractCompleted: (index) {
        print('Quick interaction selected: $index');
      },
      animate: true,
      config: QuickInteractConfig(),
      showToolTipDelay: const Duration(milliseconds: 300),
      child: const Icon(Icons.send),
    ),
  );
}
```

## API

### QuickInteractions

QuickInteractions is a widget that provides quick interaction functionality. It uses long press gesture to trigger a
set of quick interaction widgets.

#### Attributes:

- `quickInteractionWidgets`: The list of widgets for quick interactions.
- `config`: The configuration model for QuickInteractBuilder and QuickInteract classes.
- `onQuickInteractCompleted`: The callback function when a quick interaction is completed.
- `child`: The child widget.
- `onTap`: The callback function when the widget is tapped.
- `toolTip`: The tooltip widget.
- `disable`: A flag to disable the quick interactions.
- `animate`: A flag to enable animation for the quick interactions.
- `showToolTipDelay`: The delay before showing the tooltip.

### QuickInteractConfig

`QuickInteractConfig` is a configuration model for `QuickInteractBuilder` and `QuickInteract` classes. It allows you to
customize the behavior and appearance of the quick interaction widgets.
Parameters

- `widgetSize`: Size of the quick interaction widgets container. Default is 35.
- `widgetPadding`: Padding between quick interaction widgets. Default is 4.
- `transitionAnimationEndOffset`: Offset animations end value. Default is Offset(0, -1).
- `scaleAnimationEndScale`: Scale animations end value. Default is 1.5.
- `cursorAnimationDuration`: Duration of the cursor animation. Default is Duration(milliseconds: 300).
- `containerColor`: Color of the quick interaction container. No default value.
- `containerHeight`: Height of the quick interaction container. Default is 40.
- `containerBorderRadius`: Border radius of the quick interaction container. Default is BorderRadius.all(
  Radius.circular(30)).
- `containerAnimationDuration`: Duration of the container animation. Default is Duration(milliseconds: 300).
- `containerAnimationCurve`: Curve of the container animation. Default is Curves.linearToEaseOut.
- `widgetAnimationCurve`: Curve of the widget animation. Default is Curves.linearToEaseOut.
- `elevation`: Elevation of the quick interaction container. Default is 5.
  Named Constructors

- `QuickInteractConfig()`: Creates a new instance of `QuickInteractConfig` with default values.
- `QuickInteractConfig.onlyTransition()`: Creates a new instance of `QuickInteractConfig` with only transition offset
  -2. The default elevation is set to 5.
- `QuickInteractConfig.onlyScale()`: Creates a new instance of `QuickInteractConfig` with only scale 2. The default
  elevation is set to 5.

---

### Examples

You can find more examples and different
usages [in the example main.dart](https://github.com/onatcipli/quick_interact/blob/master/example/lib/main.dart).

### Upcoming Features

- Displaying quick interaction in a safe area
- Providing a builder constructor to change state when active with boolean
- For more information, please refer to the API documentation.


CHANGELOG.md
## 1.0.1

* update showcases

## 1.0.0

* initial release for [pub.dev](https://pub.dev/packages/quick_interact)

## 0.0.1

* Bulk commit


test/quick_interact_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:quick_interact/quick_interact.dart';


void main() {
  group('QuickInteractConfig', () {
    test('default values', () {
      const config = QuickInteractConfig();
      expect(config.widgetSize, 35);
      expect(config.widgetPadding, 4);
      // Add more assertions for the rest of the default values
    });
  });
}


lib/src/quick_interact.dart
part of quick_interact;

/// [QuickInteractions] is a widget that provides quick interaction functionality.
/// It uses long press gesture to trigger a set of quick interaction widgets.
///
/// Attributes:
/// - `quickInteractionWidgets`: The list of widgets for quick interactions.
/// - `config`: The configuration model for QuickInteractBuilder and QuickInteract classes.
/// - `onQuickInteractCompleted`: The callback function when a quick interaction is completed.
/// - `child`: The child widget.
/// - `onTap`: The callback function when the widget is tapped.
/// - `toolTip`: The tooltip widget.
/// - `disable`: A flag to disable the quick interactions.
/// - `animate`: A flag to enable animation for the quick interactions.
/// - `showToolTipDelay`: The delay before showing the tooltip.
class QuickInteractions extends StatefulWidget {
  const QuickInteractions({
    Key? key,
    required this.quickInteractionWidgets,
    required this.onQuickInteractCompleted,
    required this.child,
    this.onTap,
    this.toolTip,
    this.disable = false,
    this.animate = true,
    this.showToolTipDelay = const Duration(
      seconds: 2,
    ),
    this.config = const QuickInteractConfig(),
  }) : super(key: key);

  /// The delay before showing the tooltip.
  final Duration showToolTipDelay;

  /// The tooltip widget.
  final Widget? toolTip;

  /// The list of widgets for quick interactions.
  final List<Widget> quickInteractionWidgets;

  /// The callback function when the widget is tapped.
  final VoidCallback? onTap;

  /// The callback function when a quick interaction is completed.
  final Function(int) onQuickInteractCompleted;

  /// The child widget.
  final Widget child;

  /// A flag to disable the quick interactions.
  final bool disable;

  /// A flag to enable animation for the quick interactions.
  final bool animate;

  /// The configuration model for QuickInteractBuilder and QuickInteract classes.
  final QuickInteractConfig config;

  @override
  State<QuickInteractions> createState() => _QuickInteractionsState();
}

/// [_QuickInteractionsState] is the state class for [QuickInteractions].
/// It handles the gestures and animations for the quick interactions.
class _QuickInteractionsState extends State<QuickInteractions> {
  final GlobalKey _globalKey = GlobalKey();
  RenderBox? _renderObject;
  Offset? _globalPosition;
  OverlayEntry? entry;
  GlobalKey<QuickInteractContainerBuilderState>? controller;

  int? latestIndex;

  @override
  void setState(VoidCallback fn) {
    if (mounted) {
      super.setState(fn);
    }
  }

  @override
  void initState() {
    super.initState();
    if (widget.disable) {
      return;
    }
    WidgetsBinding.instance.addPostFrameCallback((timeStamp) {
      _renderObject =
          _globalKey.currentContext?.findRenderObject() as RenderBox;
      _globalPosition = _renderObject?.localToGlobal(
        Offset.zero,
      );
      if (widget.toolTip != null) {
        Future.delayed(widget.showToolTipDelay).then((value) {
          if (mounted) showTooltip();
        });
      }
    });
  }

  @override
  void dispose() {
    super.dispose();
    if (widget.disable) {
      return;
    }
  }

  @override
  Widget build(BuildContext context) {
    return GestureDetector(
      key: _globalKey,
      behavior: HitTestBehavior.opaque,
      onTap: widget.onTap ?? onTap,
      onLongPressStart: onLongPressStart,
      onLongPressCancel: onLongPressCancel,
      onLongPressEnd: onLongPressEnd,
      onLongPressMoveUpdate: onLongPressMoveUpdate,
      child: widget.child,
    );
  }

  /// Handles the long press start gesture.
  /// It shows the quick interaction widgets and activates the first one.
  void onLongPressStart(LongPressStartDetails details) {
    if (widget.disable) {
      return;
    }
    if (widget.quickInteractionWidgets.isNotEmpty) {
      show();
      controller?.currentState?.activateInteraction(0);
      Vibrate.feedback(FeedbackType.heavy);
    }
  }

  /// Handles the long press move update gesture.
  /// It animates the cursor based on the current position.
  void onLongPressMoveUpdate(LongPressMoveUpdateDetails details) {
    if (widget.disable) {
      return;
    }
    animateCursor(details.localPosition);
  }

  /// Animates the cursor based on the current position.
  /// It activates and deactivates the quick interaction widgets based on the cursor position.
  void animateCursor(Offset offset) {
    if (checkIsValidPosition(offset)) {
      if (latestIndex != null) {
        controller?.currentState?.deActivateInteraction(latestIndex!);
        latestIndex = null;
      }
      return;
    }
    final index = getIndexFromCurrentOffset(offset);
    if (latestIndex == null) {
      latestIndex = index;
      Vibrate.feedback(FeedbackType.medium);
      controller?.currentState?.activateInteraction(index);
    } else {
      if (latestIndex != index) {
        controller?.currentState?.deActivateInteraction(latestIndex!);
        controller?.currentState?.activateInteraction(index);
        Vibrate.feedback(FeedbackType.medium);
        latestIndex = index;
      }
    }
  }

  /// Handles the long press cancel gesture.
  /// It hides the quick interaction widgets with animation.
  void onLongPressCancel() {
    if (widget.disable) {
      return;
    }
    hideAnimation();
  }

  /// Handles the long press end gesture.
  /// It hides the quick interaction widgets with animation and triggers the quick interaction completed callback.
  Future<void> onLongPressEnd(LongPressEndDetails details) async {
    if (widget.disable) {
      return;
    }
    hideAnimation();
    if (checkIsValidPosition(details.localPosition)) {
      Vibrate.feedback(FeedbackType.error);
      return;
    }
    final index = getIndexFromCurrentOffset(details.localPosition);
    if (widget.animate) {
      animateInteraction(index);
    }

    /// send request
    widget.onQuickInteractCompleted(index);
    Vibrate.feedback(FeedbackType.success);
  }

  /// Gets the index of the quick interaction widget from the current offset.
  int getIndexFromCurrentOffset(Offset offset) {
    final current = ((offset.dx /
            (widget.config.widgetSize + (widget.config.widgetPadding * 2)))
        .floor());
    return current > widget.quickInteractionWidgets.length - 1
        ? widget.quickInteractionWidgets.length - 1
        : current < 0
            ? 0
            : current;
  }

  /// Shows the quick interaction widgets.
  void show() {
    if (entry != null) {
      removeWithoutAnimation();
    }
    controller = GlobalKey<QuickInteractContainerBuilderState>();
    entry = OverlayEntry(
      opaque: false,
      builder: (BuildContext context) {
        return QuickInteractContainerBuilder(
          key: controller,
          globalPosition: _globalPosition!,
          quickInteractWidgets: widget.quickInteractionWidgets,
          config: widget.config,
          removeOverlay: () => removeWithoutAnimation(),
        );
      },
    );
    Overlay.of(context).insert(entry!);
  }

  /// Animates the interaction with the given index.
  animateInteraction(int index) {
    final child = widget.quickInteractionWidgets.elementAt(index);
    late OverlayEntry entry;
    entry = OverlayEntry(
      opaque: false,
      builder: (BuildContext context) {
        return AnimateInteractWidget(
          current: _globalPosition!,
          entry: entry,
          child: child,
        );
      },
    );
    Overlay.of(context).insert(entry);
  }

  /// Animates the interaction with the given widget.
  animateInteractionWidget(Widget child) {
    late OverlayEntry entry;
    entry = OverlayEntry(
      opaque: false,
      builder: (BuildContext context) {
        return AnimateInteractWidget(
          current: _globalPosition!,
          entry: entry,
          child: child,
        );
      },
    );
    Overlay.of(context).insert(entry);
  }

  /// Hides the quick interaction widgets with animation.
  void hideAnimation() {
    if (controller != null) {
      controller!.currentState!.hideAnimation();
    }
  }

  /// Removes the quick interaction widgets without animation.
  void removeWithoutAnimation() {
    entry!.remove();
    entry = null;
    controller = null;
    latestIndex = null;
  }

  /// Handles the tap gesture.
  /// It shows the tooltip.
  void onTap() {
    if (widget.disable) {
      return;
    }
    showTooltip();
  }

  /// Checks if the given offset is a valid position for the quick interactions.
  bool checkIsValidPosition(Offset offset) {
    return offset.dy > 0 || offset.dy.abs() > 100;
  }

  /// Shows the tooltip with the given text.
  void showTooltip({String? text}) async {
    try {
      late OverlayEntry entry;
      entry = OverlayEntry(
        opaque: false,
        builder: (BuildContext context) {
          return Positioned(
            top: _globalPosition!.dy - 30,
            left: _globalPosition!.dx,
            child: Material(
              color: Colors.transparent,
              elevation: 0,
              child: widget.toolTip,
            ),
          );
        },
      );
      Overlay.of(context).insert(entry);
      await Future.delayed(const Duration(seconds: 3));
      entry.remove();
    } catch (e) {
      // print(e);
    }
  }
}


lib/src/animate_ineract_widget.dart
part of quick_interact;

/// [AnimateInteractWidget] is a widget that animates the interaction of a child widget.
/// It uses the [OverlayEntry] to display the child widget and animate it.
///
/// Attributes:
/// - [child]: The widget to animate.
/// - [current]: The current offset of the widget.
/// - [entry]: The overlay entry used to display the widget.
class AnimateInteractWidget extends StatefulWidget {
  const AnimateInteractWidget({
    Key? key,
    required this.child,
    required this.current,
    required this.entry,
  }) : super(key: key);

  final Widget child;
  final Offset current;
  final OverlayEntry entry;

  @override
  State<AnimateInteractWidget> createState() => _AnimateInteractWidgetState();
}

/// [_AnimateInteractWidgetState] is the state of [AnimateInteractWidget].
/// It handles the animation of the widget.
class _AnimateInteractWidgetState extends State<AnimateInteractWidget>
    with TickerProviderStateMixin {
  late Animation<double> dxAnimation;
  late Animation<double> opacityAnimation;
  late Animation<double> dyAnimation;
  late AnimationController dxAnimationController;
  late AnimationController dyAnimationController;

  final duration = const Duration(milliseconds: 3000);

  num swingAnimation = 25;

  @override
  void initState() {
    super.initState();
    dxAnimationController = AnimationController(
      vsync: this,
      duration: duration ~/ 4,
    );
    dyAnimationController = AnimationController(
      vsync: this,
      duration: duration,
    );
    dxAnimation = Tween<double>(
      begin: widget.current.dx,
      end: widget.current.dx + swingAnimation,
    ).animate(
      CurvedAnimation(
        parent: dxAnimationController,
        curve: Curves.easeInOut,
      ),
    );
    dyAnimation = Tween<double>(
      begin: widget.current.dy,
      end: widget.current.dy - 800,
    ).animate(
      CurvedAnimation(
        parent: dyAnimationController,
        curve: Curves.linear,
      ),
    );
    opacityAnimation = Tween<double>(
      begin: 1,
      end: 0,
    ).animate(
      CurvedAnimation(
        parent: dyAnimationController,
        curve: Curves.linear,
      ),
    );

    dyAnimationController.forward();
    dxAnimationController.repeat(
      reverse: true,
    );
    dxAnimationController.addListener(() {
      setState(() {});
    });
    dyAnimationController.addListener(() {
      setState(() {});
    });
    Future.delayed(duration).then((value) {
      widget.entry.remove();
      dxAnimationController.stop();
      dyAnimationController.stop();
    });
  }

  @override
  void dispose() {
    super.dispose();
    dxAnimationController.dispose();
    dyAnimationController.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Positioned(
      top: dyAnimation.value,
      left: dxAnimation.value,
      child: Opacity(
        opacity: opacityAnimation.value,
        child: widget.child,
      ),
    );
  }
}


lib/src/quick_interact_config_model.dart
part of quick_interact;

/// [QuickInteractConfig] is a configuration model for QuickInteractBuilder and QuickInteract classes.
/// Attributes:
/// - [widgetSize]: Size of the quick interaction widgets container.
/// - [widgetPadding]: Padding between quick interaction widgets.
/// - [transitionAnimationEndOffset]: Offset animations end value.
/// - [scaleAnimationEndScale]: Scale animations end value.
/// - [containerColor]: Color of the quick interaction container.
/// - [containerHeight]: Height of the quick interaction container.
/// - [containerAnimationDuration]: Duration of the container animation.
/// - [cursorAnimationDuration]: Duration of the cursor animation.
/// - [containerAnimationCurve]: Curve of the container animation.
/// - [widgetAnimationCurve]: Curve of the widget animation.
class QuickInteractConfig {
  /// Size of the quick interaction widgets container.
  final double widgetSize;

  /// Padding between quick interaction widgets.
  final double widgetPadding;

  /// Offset animations end value.
  final Offset transitionAnimationEndOffset;

  /// Scale animations end value.
  final double scaleAnimationEndScale;

  /// Duration of the cursor animation.
  final Duration cursorAnimationDuration;

  /// Color of the quick interaction container.
  final Color? containerColor;

  /// Height of the quick interaction container.
  final double containerHeight;

  /// Border radius of the quick interaction container.
  final BorderRadius containerBorderRadius;

  /// Duration of the container animation.
  final Duration containerAnimationDuration;

  /// Curve of the container animation.
  final Curve containerAnimationCurve;

  /// Curve of the widget animation.
  final Curve widgetAnimationCurve;

  /// Elevation of the quick interaction container.
  final double elevation;

  /// Creates a new instance of [QuickInteractConfig].
  /// The default elevation is set to 5.
  const QuickInteractConfig({
    this.widgetSize = 35,
    this.widgetPadding = 4,
    this.cursorAnimationDuration = const Duration(milliseconds: 300),
    this.containerColor,
    this.containerHeight = 40,
    this.containerBorderRadius = const BorderRadius.all(
      Radius.circular(30),
    ),
    this.containerAnimationDuration = const Duration(milliseconds: 300),
    this.containerAnimationCurve = Curves.linearToEaseOut,
    this.widgetAnimationCurve = Curves.linearToEaseOut,
    this.transitionAnimationEndOffset = const Offset(0, -1),
    this.scaleAnimationEndScale = 1.5,
    this.elevation = 5,
  }) : assert(widgetSize <= containerHeight,
            'Widget size cannot be less than container height');

  /// Creates a new instance of [QuickInteractConfig] with only transition offset -2.
  /// The default elevation is set to 5.
  const QuickInteractConfig.onlyTransition({
    this.widgetSize = 50,
    this.widgetPadding = 2,
    this.cursorAnimationDuration = const Duration(milliseconds: 300),
    this.containerColor,
    this.containerHeight = 50,
    this.containerBorderRadius = const BorderRadius.all(
      Radius.circular(30),
    ),
    this.containerAnimationDuration = const Duration(milliseconds: 300),
    this.containerAnimationCurve = Curves.linearToEaseOut,
    this.widgetAnimationCurve = Curves.linearToEaseOut,
    this.transitionAnimationEndOffset = const Offset(0, -1),
    this.scaleAnimationEndScale = 1,
    this.elevation = 5,
  }) : assert(widgetSize <= containerHeight,
            'Widget size cannot be less than container height');

  /// Creates a new instance of [QuickInteractConfig] with only scale 2.
  /// The default elevation is set to 5.
  const QuickInteractConfig.onlyScale({
    this.widgetSize = 35,
    this.widgetPadding = 5,
    this.cursorAnimationDuration = const Duration(milliseconds: 300),
    this.containerColor,
    this.containerHeight = 40,
    this.containerBorderRadius = const BorderRadius.all(
      Radius.circular(30),
    ),
    this.containerAnimationDuration = const Duration(milliseconds: 300),
    this.containerAnimationCurve = Curves.linearToEaseOut,
    this.widgetAnimationCurve = Curves.linearToEaseOut,
    this.transitionAnimationEndOffset = const Offset(0, 0),
    this.scaleAnimationEndScale = 2,
    this.elevation = 5,
  }) : assert(widgetSize <= containerHeight,
            'Widget size cannot be less than container height');
}


lib/src/quick_interact_widget_builder.dart
part of quick_interact;

/// [QuickInteractWidgetBuilder] is a widget builder for QuickInteractContainerBuilder.
/// It uses animations to create a slide and scale transition for the quick interaction widgets.
///
/// Attributes:
/// - `child`: The widget to be animated.
/// - `offsetAnimation`: The animation that controls the position of the widget.
/// - `scaleAnimation`: The animation that controls the scale of the widget.
/// - `config`: Configuration model for QuickInteractWidgetBuilder.
class QuickInteractWidgetBuilder extends StatelessWidget {
  /// Creates a [QuickInteractWidgetBuilder].
  ///
  /// The [child], [offsetAnimation], [scaleAnimation], and [config] arguments must not be null.
  const QuickInteractWidgetBuilder({
    Key? key,
    required this.child,
    required this.offsetAnimation,
    required this.scaleAnimation,
    required this.config,
  }) : super(key: key);

  /// The widget to be animated.
  final Widget child;

  /// The animation that controls the position of the widget.
  final Animation<Offset> offsetAnimation;

  /// The animation that controls the scale of the widget.
  final Animation<double> scaleAnimation;

  /// Configuration model for QuickInteractWidgetBuilder.
  final QuickInteractConfig config;

  /// Builds the widget tree.
  ///
  /// Returns a [Padding] widget that contains a [SlideTransition] widget,
  /// which in turn contains a [ScaleTransition] widget. The [ScaleTransition]
  /// widget contains a [SizedBox] that has the [child] widget.
  @override
  Widget build(BuildContext context) {
    return Padding(
      padding: EdgeInsets.all(config.widgetPadding),
      child: SlideTransition(
        position: offsetAnimation,
        child: ScaleTransition(
          scale: scaleAnimation,
          child: SizedBox(
            height: config.widgetSize,
            width: config.widgetSize,
            child: child,
          ),
        ),
      ),
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:quick_interact/quick_interact.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Quick Interact Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(),
    );
  }
}

class MyHomePage extends StatelessWidget {
  const MyHomePage({super.key});

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('Quick Interact Demo'),
      ),
      body: Padding(
        padding: const EdgeInsets.all(8.0),
        child: Column(
          mainAxisSize: MainAxisSize.max,
          mainAxisAlignment: MainAxisAlignment.spaceAround,
          children: [
            const Text(
              'long press icons to activate Quick Interact and hover and release.',
            ),
            const SizedBox(
              height: 100,
            ),
            Center(
              child: QuickInteractions(
                quickInteractionWidgets: const [
                  Icon(Icons.thumb_up),
                  Icon(Icons.thumb_down),
                  Icon(Icons.favorite),
                  Icon(Icons.bookmark),
                ],
                onQuickInteractCompleted: (index) {
                  print('Quick interaction selected: $index');
                },
                toolTip: const Text('Long press \nfor the quick interaction'),
                showToolTipDelay: const Duration(seconds: 1),
                config: const QuickInteractConfig(),
                child: const Icon(Icons.send),
              ),
            ),
            Expanded(
              child: Row(
                mainAxisAlignment: MainAxisAlignment.center,
                mainAxisSize: MainAxisSize.max,
                children: [
                  Column(
                    mainAxisAlignment: MainAxisAlignment.spaceEvenly,
                    crossAxisAlignment: CrossAxisAlignment.start,
                    children: <Widget>[
                      iconQuickReaction(
                        config: const QuickInteractConfig(),
                      ),
                      iconQuickReaction(
                        config: const QuickInteractConfig.onlyScale(),
                      ),
                      iconQuickReaction(
                        config: const QuickInteractConfig.onlyTransition(),
                      ),
                    ],
                  ),
                  const SizedBox(
                    width: 50,
                  ),
                  Column(
                    mainAxisAlignment: MainAxisAlignment.spaceEvenly,
                    children: <Widget>[
                      userAvatarQuickReaction(
                        config: const QuickInteractConfig(),
                      ),
                      userAvatarQuickReaction(
                        config: const QuickInteractConfig.onlyScale(),
                      ),
                      userAvatarQuickReaction(
                        config: const QuickInteractConfig.onlyTransition(),
                      ),
                    ],
                  ),
                ],
              ),
            ),
          ],
        ),
      ),
    );
  }

  Widget userAvatarQuickReaction({required QuickInteractConfig config}) {
    return Center(
      child: QuickInteractions(
        quickInteractionWidgets: const [
          CircleAvatar(
            radius: 15,
            backgroundImage:
                NetworkImage('https://randomuser.me/api/portraits/men/1.jpg'),
          ),
          CircleAvatar(
            radius: 15,
            backgroundImage:
                NetworkImage('https://randomuser.me/api/portraits/women/2.jpg'),
          ),
          CircleAvatar(
            radius: 15,
            backgroundImage:
                NetworkImage('https://randomuser.me/api/portraits/men/3.jpg'),
          ),
          CircleAvatar(
            radius: 15,
            backgroundImage:
                NetworkImage('https://randomuser.me/api/portraits/women/4.jpg'),
          ),
        ],
        onQuickInteractCompleted: (index) {
          print('Quick interaction selected: $index');
        },
        animate: true,
        config: config,
        showToolTipDelay: const Duration(milliseconds: 300),
        child: const Icon(Icons.send),
      ),
    );
  }

  Widget iconQuickReaction({required QuickInteractConfig config}) {
    return QuickInteractions(
      quickInteractionWidgets: const [
        Icon(Icons.thumb_up),
        Icon(Icons.thumb_down),
        Icon(Icons.favorite),
        Icon(Icons.bookmark),
      ],
      onQuickInteractCompleted: (index) {
        print('Quick interaction selected: $index');
      },
      config: config,
      child: const Icon(Icons.send),
    );
  }
}
